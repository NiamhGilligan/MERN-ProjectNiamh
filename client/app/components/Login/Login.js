import React, { Component } from 'react';
import 'whatwg-fetch';

import MyAccount from '../AccountHomePage/MyAccountPage';
import {
  getFromStorage,
  setInStorage,
} from "../../utils/storage";


class Login extends Component {
  constructor(props){
    super(props);
    this.state ={
      token:'',
      signInError:'',
      signInEmail: '',
      signInPassword:'',
      valid: false,
    };
      this.onTextboxChangeSignInEmail = this.onTextboxChangeSignInEmail.bind(this);
    this.onTextboxChangeSignInPassword = this.onTextboxChangeSignInPassword.bind(this);
    this.onSignIn = this.onSignIn.bind(this);
    this.logout = this.logout.bind(this);

  }


  componentDidMount() {
    this.forceUpdate();
    const obj = getFromStorage('the_main_app');
    if(obj && obj.token){
      const {token} = obj;
      //verifying token
      fetch('/api/account/verify?token=' + token)
        .then(res => res.json())
        .then(json =>{
          if(json.success){
            this.setState({
              token,
              isLoading:false
            });
          }else{
            this.setState({
              isLoading:false,
            });
          }
        });
    }else{
      this.setState({
        isLoading:false,
      });
    }
  }


  onTextboxChangeSignInEmail(event){
    this.setState({
      signInEmail:event.target.value,
    });
  }
  onTextboxChangeSignInPassword(event){
    this.setState({

      signInPassword:event.target.value
    });
  }

  onSignIn(){
    //grabbing the state and post request to backend
    const {

      signInEmail,
      signInPassword,

    } = this.state;
    this.setState({
      isLoading:true,
    });
    fetch('/api/account/signin', {
      method:'POST',
      headers:{
        'Content-Type':'application/json'
      },
      body:JSON.stringify({
        email: signInEmail,
        password:signInPassword,
      }),
    }).then(res => res.json())
      .then(json => {

        if (!json.success){
          alert('Incorrect Details. Please try again');
          window.location.reload()
        }
        if (json.success){
          setInStorage('the_main_app',{token: json.token});
          setInStorage('user', json.user._id);
          this.setState({
            signInError: json.message,
            isLoading: false,
            signInEmail: '',
            signInPassword:'',
            token:json.token,
          });

        }


      });
  }
  logout(){
    this.setState({
      isLoading:true,
    });
    const obj = getFromStorage('the_main_app');
    if(obj && obj.token){
      const {token} = obj;
      //verifying token
      fetch('/api/account/logout?token=' + token)
        .then(res => res.json())
        .then(json =>{
          if(json.success){
            this.setState({
              token:'',
              isLoading:false
            });
          }else{
            this.setState({
              isLoading:false,
            });
          }
        });
    }else{
      this.setState({
        isLoading:false,
      });
    }
  }


  render() {
    const {
      isLoading,
      token,
      signInError,
      signInEmail,
      signInPassword,
      valid,

    } = this.state;

    if (isLoading) {
      return (<div><p>Loading....</p></div>);
    }
    if (!token) {
      return (

        <div className="loginBox">
          <button
               className="user fa fa-address-card btn-default"> <br/>Teachers Toolbox</button>
          <h2>Log in Here</h2>
            <br/>
            <p>Email</p>
            <input
              className="valinput"
              name="email"
              placeholder='email'
              value={signInEmail}
              onChange={this.onTextboxChangeSignInEmail}
              type="text"
              id="inputUsername"
              required
            />
          <small>{this.email ? this.signInEmail.validationMessage : null}</small>
            <br/>
            <p>Password</p>
            <input
              id ="password"
              className="valinput"
              name="password"
              type="password"
              placeholder='**********'
              value={signInPassword}
              onChange={this.onTextboxChangeSignInPassword}
              minLength={4}
              required
            />
          <small>{this.password ? this.signInPassword.validationMessage : null}</small>
            <br/><br/>

            <button
              onClick={this.onSignIn}>Submit</button>
            <br/>
            <a href="#">Forgot your password ?</a>


        </div>

      );
    }
    return (
      <div>
        <div className="alert alert-success">
          <strong>Success!</strong> You are logged in!.
          <button className ="btn-success" onClick={this.logout}>Logout</button>
        </div>

        <MyAccount/>

      </div>
    );
  }
}



export default Login;
