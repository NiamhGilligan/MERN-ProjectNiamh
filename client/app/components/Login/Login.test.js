import React from 'react';
import {shallow} from 'enzyme';
import Login from './Login';

describe('Login Test', ()=> {

  test('Rendering Login page',() => {
const wrapper = shallow(
  <Login/>
);

expect(wrapper).toMatchSnapshot();
  });

  it('renders a email input', () => {
    expect(shallow(<Login />).find('#inputUsername').length).toEqual(1)
  });
  it('renders a password input', () => {
    expect(shallow(<Login />).find('#password').length).toEqual(1)
  });




});
