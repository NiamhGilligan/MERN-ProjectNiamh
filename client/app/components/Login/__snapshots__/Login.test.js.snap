// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Test Rendering Login page 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Login />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <button
          className="user fa fa-address-card btn-default"
        >
           
          <br />
          Teachers Toolbox
        </button>,
        <h2>
          Log in Here
        </h2>,
        <br />,
        <p>
          Email
        </p>,
        <input
          className="valinput"
          id="inputUsername"
          name="email"
          onChange={[Function]}
          placeholder="email"
          required={true}
          type="text"
          value=""
        />,
        <small />,
        <br />,
        <p>
          Password
        </p>,
        <input
          className="valinput"
          id="password"
          minLength={4}
          name="password"
          onChange={[Function]}
          placeholder="**********"
          required={true}
          type="password"
          value=""
        />,
        <small />,
        <br />,
        <br />,
        <button
          onClick={[Function]}
        >
          Submit
        </button>,
        <br />,
        <a
          href="#"
        >
          Forgot your password ?
        </a>,
      ],
      "className": "loginBox",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            " ",
            <br />,
            "Teachers Toolbox",
          ],
          "className": "user fa fa-address-card btn-default",
        },
        "ref": null,
        "rendered": Array [
          " ",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": "br",
          },
          "Teachers Toolbox",
        ],
        "type": "button",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Log in Here",
        },
        "ref": null,
        "rendered": "Log in Here",
        "type": "h2",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": "br",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Email",
        },
        "ref": null,
        "rendered": "Email",
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "valinput",
          "id": "inputUsername",
          "name": "email",
          "onChange": [Function],
          "placeholder": "email",
          "required": true,
          "type": "text",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": null,
        },
        "ref": null,
        "rendered": null,
        "type": "small",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": "br",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Password",
        },
        "ref": null,
        "rendered": "Password",
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "valinput",
          "id": "password",
          "minLength": 4,
          "name": "password",
          "onChange": [Function],
          "placeholder": "**********",
          "required": true,
          "type": "password",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": null,
        },
        "ref": null,
        "rendered": null,
        "type": "small",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": "br",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": "br",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Submit",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Submit",
        "type": "button",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": "br",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Forgot your password ?",
          "href": "#",
        },
        "ref": null,
        "rendered": "Forgot your password ?",
        "type": "a",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <button
            className="user fa fa-address-card btn-default"
          >
             
            <br />
            Teachers Toolbox
          </button>,
          <h2>
            Log in Here
          </h2>,
          <br />,
          <p>
            Email
          </p>,
          <input
            className="valinput"
            id="inputUsername"
            name="email"
            onChange={[Function]}
            placeholder="email"
            required={true}
            type="text"
            value=""
          />,
          <small />,
          <br />,
          <p>
            Password
          </p>,
          <input
            className="valinput"
            id="password"
            minLength={4}
            name="password"
            onChange={[Function]}
            placeholder="**********"
            required={true}
            type="password"
            value=""
          />,
          <small />,
          <br />,
          <br />,
          <button
            onClick={[Function]}
          >
            Submit
          </button>,
          <br />,
          <a
            href="#"
          >
            Forgot your password ?
          </a>,
        ],
        "className": "loginBox",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              " ",
              <br />,
              "Teachers Toolbox",
            ],
            "className": "user fa fa-address-card btn-default",
          },
          "ref": null,
          "rendered": Array [
            " ",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": "br",
            },
            "Teachers Toolbox",
          ],
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Log in Here",
          },
          "ref": null,
          "rendered": "Log in Here",
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "br",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Email",
          },
          "ref": null,
          "rendered": "Email",
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "valinput",
            "id": "inputUsername",
            "name": "email",
            "onChange": [Function],
            "placeholder": "email",
            "required": true,
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": null,
          },
          "ref": null,
          "rendered": null,
          "type": "small",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "br",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Password",
          },
          "ref": null,
          "rendered": "Password",
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "valinput",
            "id": "password",
            "minLength": 4,
            "name": "password",
            "onChange": [Function],
            "placeholder": "**********",
            "required": true,
            "type": "password",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": null,
          },
          "ref": null,
          "rendered": null,
          "type": "small",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "br",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "br",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Submit",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Submit",
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "br",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Forgot your password ?",
            "href": "#",
          },
          "ref": null,
          "rendered": "Forgot your password ?",
          "type": "a",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
}
`;
